### DISCUSSION TITLE
Mesh read performace enhancement with Ensight reader

### DISCUSSION THREAD

**Discussion by AliE89** [pyvista/pyvista#3802] [Q&A]
Hello pyVISTA devs and community!

I am finally trying out this great piece of software and I am here to have some advices (I am a noob! :p)

Today I have tried to load a big dataset from a CFD simulation of around 80Millions cells, saved in ensight format. It took 25 mins to load and doing an orthogonal slice takes around 2min.  Am I doing something bad or am I missing some important configurations? Any chance I can do better than this?

Thank you and sorry for the silly question!

**Comment by MatthewFlamm** [comment]
This an exceptionally long time to read that amount of data in my experience, although I only have limited experience with Ensight data format.  It is also going to heavily depend on hardware.  Is it binary or ASCII data?  Are you loading data into swap memory?

Any mesh operation on data sets this large will take some time to run.  It will be affected by type of dataset (Im assuming you have an UnstructuredGrid), number of cell and point data and possibly the complexity of the mesh.  Also again your hardware.

If you do not need all the point and cell data, you can see if not reading in that data helps, particularly if you have a large number of those. For example try turning off all point and cell data in the reader and see if the performance improves.  If so, only read in the ones that you need.


```py
import pyvista as pv
reader = pv.get_reader(<path_to_ensight_file>)
reader.disable_all_cell_arrays()
reader.disable_all_point_arrays()
mesh = reader.read()
```

**Reply by AliE89** [reply]
Cheers, I'll try this out. The ensight file is indeed ascii and not binary... I'll try to save a binary one if I can and get you back again! The dataset is unstructured grid  as you say.

**Reply by AliE89** [reply]
So... I am getting this in binary.  The file is produce by the ensight gold exporter by Ansys Fluent.
![image](https://user-images.githubusercontent.com/49912570/211510766-44f1e564-7e5d-4bdb-a8e9-31ace6deb965.png)

**Reply by banesullivan** [reply]
cc @akaszynski, this file is coming from Fluent

**Reply by MatthewFlamm** [reply]
The Ensight gold format I have handy from Ansys Fluent is only 1+ million elements and 7 cell arrays. It is in binary format.  It takes 712 ms to read on my machine using Pyvista 0.37.0 and vtk 9.1.0.  Disabling all cell data only decreases the read time to 662 ms.  If I multiply by 80, we might expect a few minutes of read time, assuming the read time is proportional to mesh size.

**Reply by MatthewFlamm** [reply]
I mocked up a simple (100x100x500) hexahedral mesh and read into Fluent v2022r2 and initialized a case and then output in binary and ascii format.  I can no longer read either of them into pyvista with errors like `Unable to open file "ensight_data.geo"`.  So I wonder if this is a change in Fluent and/or EnSight that hasn't been updated in vtk?

Edit: my working copy was from an older version of Fluent (19.1 maybe?)

More edit: I have same problem in paraview, so I don't think this is a pyvista specific issue.

**Reply by MatthewFlamm** [reply]
I tried exporting in CGNS, and it reads into pyvista in just a few seconds.  You can try this too.

**Reply by bndunn** [reply]
Matthew,
Please run the application included with ensight, called ens_checker on your 100x100x500 hex mesh's .case file _in the same directory as your case _file__
e.g.
ens_checker ensight_data.case
This will check the format of the case file and the geometry file (.geo) file and all the variable files, and echo back in detail what is wrong if it is a formatting issue.

**Reply by MatthewFlamm** [reply]
I'm not sure about @AliE89, but I have used this methodology as an intermediate step to read Fluent data into vtk/pyvista.  Fluent data in the newer releases are not supported by `vtkFluentReader` anymore unless there is a workaround.  I do not have a license for EnSight directly, so I presume I cannot use this checker. In previous versions of Fluent, I was able to export from Fluent into EnSight which had good support for multizone output from Fluent. Edit: Then use this export to read into vtk.

In this comment https://github.com/pyvista/pyvista/discussions/3802#discussioncomment-4657263, I found that Fluent is adding double quotes and these are required to remove to be read by vtk, but I'm sure if this is a Fluent problem or (likely) a vtk problem.

**Comment by banesullivan** [comment]
Would you be able to share this specific data file?

**Reply by AliE89** [reply]
I would love to do so... but unfortunately I cannot :/

**Comment by AliE89** [comment]
Hi Metthew,

Yes, your calculations sound very reasonable! There is a chance that the
ascii format is killing performance here, but cannot get a binary file
working from that damn fluent at the moment...

Il Mar 10 Gen 2023, 16:24 MatthewFlamm ***@***.***> ha
scritto:

**Comment by AliE89** [comment]
Good question, I have emailed the support but for them it's all OK. They
told me they can load the binary file into their ensight software so that
is the end of the problem for them. 

I can load up the ascii but not the binary, so I am half way apparently.

**Reply by akaszynski** [reply]
@AliE89, please email me at alexander.kaszynski@ansys.com

I'd like to see how we can help you out.

**Reply by AliE89** [reply]
Thanks, as a disclaimer I do not really want move the mountains for this issues, I can find a workaround worst case scenario :)

**Reply by akaszynski** [reply]
No issue (or pressure) at all. I'd like to be aware of what the issue is. Hopefully there's a workaround as I'm a fan of VTK/ParaView and Ansys :)

**Reply by ahernsean** [reply]
I'm the head of EnSight product development and am also a huge fan of VTK. 🙂 I'd like to get to the bottom of what the actual problem is, too. I'll have one of my developers poke around and see what might be going on.

**Reply by bndunn** [reply]
EnSight includes a checker application with the install.
Looks like, from your error message, that your case filename is Volume.case
From a command line, _in the directory where the case file is _located...._
Try_
ens_checker Volume.case
The output to the console from this format checker will explicitly point out what it is finding wrong.
It is more severe of a checker than EnSight, so it may echo a number of warnings that are non fatal, but you should get an error if EnSight fails to read this data....

**Comment by AliE89** [comment]
Cool, which options did you use for CGNS? Does it work for unstructured
data sets too?

**Reply by MatthewFlamm** [reply]
It output an unstructured mesh, so yes.  The only options are node vs cell centered and file format.  The default HDF5 format worked for me.

**Comment by AliE89** [comment]
Thank you very much, VIP support today 😆

Il Mer 11 Gen 2023, 00:49 Sean Ahern ***@***.***> ha scritto:

**Comment by MatthewFlamm** [comment]
I think this may be related to this report:
https://discourse.paraview.org/t/fluent-2021-r1-ensight-case-gold-solution-files-to-paraview-5-9-0/7512/2

and this issue:
https://gitlab.kitware.com/paraview/paraview/-/issues/20819

**Comment by AliE89** [comment]
Hi, I am aware of that issue, but getting into troubles even deleting quote
:)

Il Mer 11 Gen 2023, 15:57 MatthewFlamm ***@***.***> ha
scritto:

**Reply by cquammen** [reply]
The double quotes and Script fix should be in VTK's `master` branch now. The fix is in https://gitlab.kitware.com/vtk/vtk/-/merge_requests/9975

**Reply by banesullivan** [reply]
And FYI, you can use these configurations to clone VTK and build a wheel: https://github.com/banesullivan/vtk-cmake
